src/lib.rs

```diff
@@ -162,13 +162,41 @@ impl<C: ManageConnection + Send> Pool<C> {
     /// This **does not** implement any timeout functionality. Timeout functionality can be added
     /// by calling `.timeout` on the returned future.
     pub async fn connection(&self) -> Result<Conn<C>, Error<C::Error>> {
+        loop {
+            let mut connection = self.connection_inner().await?;
+
+            match self.conn_pool.is_valid(&mut connection).await {
+                Ok(()) => return Ok(connection),
+                Err(e) => {
+                    debug!(
+                        "connection: found connection in pool that is no longer valid - removing from pool: {:?}",
+                        e
+                    );
+
+                    connection.forget();
+
+                    self.conn_pool.conns.decrement();
```

Comment:
Can we add a test that makes sure we never falsely exhaust connections in the pool? IE if the pool cap is 3, we should be able to check out a 4th after getting 3 bad connections.

Reply:
done